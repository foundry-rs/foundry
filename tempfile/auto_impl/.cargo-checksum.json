{"files":{"Cargo.lock":"78a0fe9b44d3aa6fffdda50ace0218e1acb34c71e3da65daf9dc61a7738973a1","Cargo.toml":"3202d4573f864683df9c0704003d0e252bec41de53322d451fd9dbb07e7bbbba","LICENSE-APACHE":"8173d5c29b4f956d532781d2b86e4e30f83e6b7878dce18c919451d6ba707c90","LICENSE-MIT":"aecd978992e074f652c672cc628d9f37846ac9ffdc5365c7c5b31e8d89f7a621","README.md":"ba2fec4b58639c0d6c22298efae441face2f4a6eb30c1f0be870687cda49af92","examples/README.md":"411834270fa90127b9886b016ef0bfe2091fc6090b98e1e89a054074ad97c566","examples/error_messages.rs":"d143dd343b02c7db57f026be9a44a7eef435626400d3db22a0c758410f593929","examples/greet_closure.rs":"f2e5c50c403c8c4a49207a1b6d18eea4d85e1b63a3249a94112a75d6067b0725","examples/keep_default_for.rs":"10b6bba633321993c7f58b6ffb07225d273297cc60d16bdb30b42cb70a71e6ae","examples/names.rs":"728824f8b5b2aec847dbe611c7d2bdd73f228a50cfe3e08b7bfe79d33eb33184","examples/refs.rs":"010a3437300753ca2964f09bc18659b3247531aa5592d2e00556c029d568620e","rustfmt.toml":"1f1b377b24d1bd6d0632e3316dcaddf33584fe01569c765952856686da94a7e8","src/analyze.rs":"84a2145848e4ed281239aaf60fc69b48a45d6d9007675ed22ed0a80ceb3885fb","src/attr.rs":"1e291173d2e38fd2e8e2ed752254dc138823c6b94b8f970077eed6766ee6cdd9","src/gen.rs":"a42ac88a21448a266f04a20d8dec6eb42e1a4ccc47d2de265744d629c49ad124","src/lib.rs":"89bc672b587ce86c1c2c6b66d16a3fbb50d1ebadee39e358519f56a5d215ddad","src/proxy.rs":"f0157a64aaad5be8fbb885104e6a85b968cfcf91bbeda8a8d77c09628881a9c1","tests/compile-fail/attr_on_enum.rs":"58aa7e6d2569eb4e829be28a80c5f50310d0c33ec00e3e1ce77c3a95291be2e5","tests/compile-fail/attr_on_enum.stderr":"71fc710f2dffc2647af1d4ab926cdf6c1699c9389d39343a683b28d70fc585ce","tests/compile-fail/attr_on_fn.rs":"af65b3f0d78dc1ac0d7efc1b6f507aa17235bcfb23a76cdcbac61f2a9c15f4f9","tests/compile-fail/attr_on_fn.stderr":"79a4e3cb997e5178421653f91261714552ff4599ffff73fddebbd0a077b470e9","tests/compile-fail/attr_on_impl_block.rs":"04630b9760400ce4245a29f68841d2fced15ed7466f20daee1fa589c35c4b7ef","tests/compile-fail/attr_on_impl_block.stderr":"7b85b05d8aa1a701f8d033b198118b18a4f1aef4b69db852bc05849826b6bde0","tests/compile-fail/attr_on_struct.rs":"91c2406cc4b3c2adab668eb7de255be185581743ae7b177cb1b0fba2415baa76","tests/compile-fail/attr_on_struct.stderr":"b57d1e2fb29901ef1fc54462865a8aba0343cf111f82b80932c9d297f4a38d47","tests/compile-fail/attr_on_type.rs":"8bb41b4100a22dda0778c9e753464f155dc3bb7e7aac0c9621fc8dd7e2cf99a6","tests/compile-fail/attr_on_type.stderr":"ae1d23de9d0c1ce0f21844c9f1ced76ea8dba1b8d66ae40af57c276a6f69c00e","tests/compile-fail/attr_on_unit_struct.rs":"35455b758bf969d60d15fad06cfa6147393d73d62c83013eca129f21538b6d89","tests/compile-fail/attr_on_unit_struct.stderr":"5e3b60179c1b97f7d3cbe799842d22251480c6ffc52ce684b09d315531747c25","tests/compile-fail/fn_associated_const.rs":"966448e0a42584443bc9a5f0eb5319a48115cb9f1b699747d1ce851161e7a575","tests/compile-fail/fn_associated_const.stderr":"7060507cab890bc9af2febc2b51a2688c17f6d28997e49f684b20d6b26adc43f","tests/compile-fail/fn_associated_type.rs":"9715638b19f5f3f4b0231bf789f9bc696993d6f8d8a39c66e524fac3516f45c1","tests/compile-fail/fn_associated_type.stderr":"7008560624ffdede5be901e8ea99bf4b5e0789f2b4d204d542428ff7c1cc240e","tests/compile-fail/fn_const_generics.rs":"1978129ed57841ef4c514d42a5b993ca99f80779143bcd9059d356e7922d299e","tests/compile-fail/fn_const_generics.stderr":"923b62298579e5f54fff0ad691b9e1e718f6f418c1acf91b0628ec844f76923e","tests/compile-fail/fn_generics.rs":"abbdc20b75fa1ec8718c2073a5140472afc3381e937379411faf9ac8019759c2","tests/compile-fail/fn_generics.stderr":"3d2e2c2fe070713aea2c58b577a56528222aca08a6d8f06e468c75a15076fb11","tests/compile-fail/fn_multiple_methods.rs":"4341c308751a46bbce050e816a7ab4a3235505ea6c4ffbccf34f817def06ff1b","tests/compile-fail/fn_multiple_methods.stderr":"5363eacc42baacdfe986e24ee7dd7a24514ab4c0500c4024e604efff0dc1c66f","tests/compile-fail/fn_unsafe_method.rs":"0cd427cbfd5c7943a022f05dbe4d218b7811c12d284081813911d143f3386e1d","tests/compile-fail/fn_unsafe_method.stderr":"e10804de046c39630f7e1b4e586a49f87379062dd642a6e62f2a81e71550fd50","tests/compile-fail/keep_default_for_on_assoc_type.rs":"a3175bef5509957a79ba4cc6a8fef0f4e0e879c0040e0b64bcab5bcbf1e0b9b4","tests/compile-fail/keep_default_for_on_assoc_type.stderr":"47f02e63338ab7138cfc23ba89c8b34571ceb4837239de49e03f0696604a7744","tests/compile-fail/keep_default_for_on_required_method.rs":"54c47d1627a97a15bd638535f17b6a20076262a0d0153eb625858483a4f85387","tests/compile-fail/keep_default_for_on_required_method.stderr":"81334f78e14236359d548b8c201c42adf2aef43a44c7b91bfb926588b9f81edc","tests/compile-fail/method_attr_invalid.rs":"573e960d96500a103cb5b103b6a8da42af7070539ff9bd05951255efbb126f6c","tests/compile-fail/method_attr_invalid.stderr":"47c65368be9889d4f5ed6882746334af1e69138228365ee48b950a8546ba7867","tests/compile-fail/mut_self_for_arc.rs":"7511117a7ed31c0eebc7a4bda8bc7d91af969135e0c0ee322cbe173b93f81ba5","tests/compile-fail/mut_self_for_arc.stderr":"b6136477b7f9edd7288f8505826baeb2e3872ad43250ad023d082b715c2a67ff","tests/compile-fail/mut_self_for_immutable_ref.rs":"2b9a4fdb2ed9b1a7ade0ec319195fbf25f1f0425c1ad154975ce5169bca1849f","tests/compile-fail/mut_self_for_immutable_ref.stderr":"8c3368e031611803859b3b2cfe2d99e9a000789632ed263acbf72e01186532ae","tests/compile-fail/mut_self_for_rc.rs":"eaaaf5df6749a5565238c86b09ae82a1b720d7899a845319e8cf9d2600d77469","tests/compile-fail/mut_self_for_rc.stderr":"49c9c9db4ed152ebd7a066014ac33ce66dadbf90b0d9e9738765afee53f9d190","tests/compile-fail/super_trait_not_implemented.rs":"b53934bc7c999c4fed49cd5e30fc1861f399d28d97d5262dde5c7bcc6565207c","tests/compile-fail/super_trait_not_implemented.stderr":"d7dae12309ddcc9917de8a40bfdccc3dbb1de91b87500ed338608b4aacea1abf","tests/compile-fail/trait_obj_value_self.rs":"4407cb9320889fb54722443506ad85f3d20446fbe41d37c5e0fe420a1e31e81a","tests/compile-fail/trait_obj_value_self.stderr":"d58b04318c38b4e97e43ead6f42e6a2636136252945c37d912f77e4888f5a9b1","tests/compile-fail/value_self_for_immutable_ref.rs":"640c014ab7cee853969e982bafe8c0ab4425db491249c9a8239c14d10d90e82a","tests/compile-fail/value_self_for_immutable_ref.stderr":"ba797ea066f1a9ef902b7cfd5bf7eb70d15b8bfd74f6313d6091be628063c129","tests/compile-fail/value_self_for_mutable_ref.rs":"a6c669f73d5b0bd47acf5b576d04b4d4013bcff8a7f164adcaddebe3bd1dee5d","tests/compile-fail/value_self_for_mutable_ref.stderr":"bfc4c0cb72f8ae32ee477fd3383588a8bb9014a8568c40e55334ad923e90de53","tests/compile-pass/big_trait_for_all_refs.rs":"2173a271900511bbabf3e9086b631314672ac60939b2255d519fed9ada02cd30","tests/compile-pass/big_trait_for_box.rs":"b8590a7daea5ae93c4fd199ddba55bef63eede5681bbebd43e3fe3c23835e94f","tests/compile-pass/big_trait_where_bound_for_all_refs.rs":"00009c98de7cd39f285bbbe773864720286645b1d6fc487fadab31d33c45c04d","tests/compile-pass/fn_method_lifetimes.rs":"eccb3ede6e38ebc42b0011d57224a96a01c21921494a9ff7e9bfd89a61b03099","tests/compile-pass/generic_fn_method_for_refs.rs":"817aa515edc2ed43caec936596e1f334949a80e382235048e2e709f62e609cdf","tests/compile-pass/generic_types_and_lifetimes_for_fn.rs":"e0209e363c99035f3150602650c101005d4966c0125d42f891c3cf2fe4905478","tests/compile-pass/generic_types_and_lifetimes_for_refs.rs":"eb29f417f5d32419849070b49f6b36ff8a0a2a92a30d90c6b4ab0fad539f4ca7","tests/compile-pass/keep_default_for_simple.rs":"2bf56940218a0c8d90bf7371a3fb36e309a050adbe873e59c82ca58d6f1f670c","tests/compile-pass/keep_default_for_with_where_bounds.rs":"edad650a239fac37679edb5443902a6a8c37bfa89686918033d73fe5dca2a06e","tests/compile-pass/method_name_shadowing.rs":"334b6b3a8b7a93516a9140a4f313f3193fdec8f0bbe19f43efc95c6e62a26e31","tests/compile-pass/mut_self_for_fn.rs":"0dcec787f22799ac9dc03414fb7119f51a559f6db69b692b7cdfbaaa3b1003fc","tests/compile-pass/mut_self_for_fn_mut.rs":"c4d9a2d155c74e0c0690f2d68825a5b50589fb0bfa16e33e0db374de3890683b","tests/compile-pass/non_inferred_generic_types_for_all_refs.rs":"7a29e9af7a832d8935cf7e89fd6dda8ccbb5db39af29dc517726407ab2df3660","tests/compile-pass/non_inferred_generic_types_for_box.rs":"f7be3dd99f1d360bce6afb496963bffc45c4016c555c281709381f1f710fd88b","tests/compile-pass/non_inferred_generic_types_with_lifetimes_for_all_refs.rs":"da7179e5efda9c6267dcc9a23e4319b2cdfaf4885a812488df5a93d589b39080","tests/compile-pass/ref_self_for_fn.rs":"2c230a8e7f32b6d8e706009fbaae0a7e4b73fcb8c5e9ea343a41c053716f2d6d","tests/compile-pass/self_bound.rs":"8aa36c8885020fafa272e5258eadda5e6673a90ed35dc48a76b4eb9b5db98148","tests/compile-pass/self_bound_default_method.rs":"ee6ac982694e506f5da99fd2a5d6130ea50e1c74f5c669105afc44bf23c304ea","tests/compile-pass/self_bound_multiple.rs":"ad7c3464573501d77483901148fa36db1d52b30991b9da21364d1cfbef10ac0d","tests/compile-pass/self_by_value_mut.rs":"55e588e47c0cc4125bc7c76f7532a5aec144746b18b8c38fe324cbc9daf461b9","tests/compile-pass/super_trait_complex_for_refs.rs":"c4e54fc0bc9d255428755039ee7453d201c5d3579f503adda11fafc01ff2870e","tests/compile-pass/super_trait_simple_for_fn_types.rs":"5f51998f8105fb9da8e61eb4e5db481756025ae52486a51f713dd9c0d73c170a","tests/compile-pass/super_trait_simple_for_refs.rs":"54d1cf31ce2f6360c899a2f46f6a233b549a1bef680b1f738535a8f15753c199","tests/compile-pass/trailing_comma.rs":"43489b6132cb374cf75444711512cf2a5926fc62150dfecfcf05c35a91c62de4","tests/compile-pass/trait_in_fn.rs":"047fa5aa51782cd7a77cece20858fa0d7e45d9a864bd930cb1c5c0b9a527dc74","tests/compile-pass/trait_in_mods.rs":"651445bb53a81fccc630deeca888175f9c13d9aaffded4ccf5f49c9d7404f952","tests/compile-pass/trait_obj_default_method.rs":"b67c943b7a4e99b0260c1d6e95d44b3e49a344b0a6fc2696374daace2da46144","tests/compile-pass/trait_obj_immutable_self.rs":"e59d38a84a15919da61fe4c834bbdface2c1f48fb7924c8440a14a3609a606d5","tests/compile-pass/trait_obj_self_sized.rs":"bf3a277328dca95a6c0566a8b33422af7757ce8dfd2b5968753f632c3888b854","tests/compile-pass/trait_obj_value_self.rs":"76bc967f51211538b9ceddbe3a3f50fe539cf502eee2bc36d803b06c8f17df28","tests/compile-pass/value_self_for_fn.rs":"a3a1ddc15b31cee0a6ce71354f3091ba528a3013e61f6f7a08b7c4e394048c75","tests/compile-pass/value_self_for_fn_mut.rs":"30eb99d725e33d00c6d1bc49d83680edfa60ff7b47793572bc85f0403f3b89ea","tests/compile-pass/value_self_for_fn_once.rs":"b5f1187a51fd08417545bf548940ad5b4ac5b4fe3c08da0630015ee81d83c68c","tests/no_std.rs":"bee43a3deb027bf1b8bfb77805f6056338b86b6666f100961b5cc859e1d146eb","tests/since_1.51/compile-pass/associated_lifetime.rs":"2f9b664cf53316d507ad0362aeebe81133e3cc425517ed2b9ca0d44f9de75b22","tests/since_1.51/compile-pass/generic_const_method.rs":"ae1d17a9915f51d7afcc1f9edb98c141886296bd593230457cbaa08e7004d945","tests/since_1.51/compile-pass/generic_mix.rs":"554b4aafb93888409982d21d48889feda59e1dd81cab8d6dce7bd7ea5de25ef9","tests/since_1.75/compile-pass/async_trait.rs":"32a4d5576b2810602f29ce638cb6603814a20704f444556a1f69b66c1a76c67d","tests/ui.rs":"64051b47b202b24e8f33a1bd5b06c30ba52d5571e2a7b1d3b8b5716fd4ac7bff"},"package":"e12882f59de5360c748c4cbf569a042d5fb0eb515f7bea9c1f470b47f6ffbd73"}