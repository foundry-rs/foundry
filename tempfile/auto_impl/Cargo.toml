# THIS FILE IS AUTOMATICALLY GENERATED BY CARGO
#
# When uploading crates to the registry Cargo will automatically
# "normalize" Cargo.toml files for maximal compatibility
# with all versions of Cargo and also rewrite `path` dependencies
# to registry (e.g., crates.io) dependencies.
#
# If you are reading this file be aware that the original Cargo.toml
# will likely look very different (and much more reasonable).
# See Cargo.toml.orig for the original contents.

[package]
edition = "2021"
rust-version = "1.56"
name = "auto_impl"
version = "1.2.1"
authors = [
    "Ashley Mannix <ashleymannix@live.com.au>",
    "Lukas Kalbertodt <lukas.kalbertodt@gmail.com>",
]
build = false
autolib = false
autobins = false
autoexamples = false
autotests = false
autobenches = false
description = "Automatically implement traits for common smart pointers and closures"
documentation = "https://docs.rs/auto_impl/"
readme = "README.md"
keywords = [
    "trait",
    "impl",
    "proc-macro",
    "closure",
]
categories = [
    "development-tools",
    "rust-patterns",
]
license = "MIT OR Apache-2.0"
repository = "https://github.com/auto-impl-rs/auto_impl/"

[lib]
name = "auto_impl"
path = "src/lib.rs"
proc-macro = true

[[example]]
name = "error_messages"
path = "examples/error_messages.rs"

[[example]]
name = "greet_closure"
path = "examples/greet_closure.rs"

[[example]]
name = "keep_default_for"
path = "examples/keep_default_for.rs"

[[example]]
name = "names"
path = "examples/names.rs"

[[example]]
name = "refs"
path = "examples/refs.rs"

[[test]]
name = "no_std"
path = "tests/no_std.rs"

[[test]]
name = "ui"
path = "tests/ui.rs"

[dependencies.proc-macro2]
version = "1.0"

[dependencies.quote]
version = "1.0"

[dependencies.syn]
version = "2.0"
features = [
    "full",
    "visit",
    "visit-mut",
]

[dev-dependencies.rustversion]
version = "1"

[dev-dependencies.trybuild]
version = "1"
