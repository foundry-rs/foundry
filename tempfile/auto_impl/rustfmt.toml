unstable_features = true

blank_lines_upper_bound = 3
format_macro_matchers = true
imports_granularity="Crate"
normalize_comments = true
reorder_impl_items = true
