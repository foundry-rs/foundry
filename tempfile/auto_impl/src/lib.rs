//! A proc-macro attribute for automatically implementing a trait for
//! references, some common smart pointers and closures.
//!
//! ## Simple example
//!
//! ```
//! use auto_impl::auto_impl;
//!
//! // This will generate two additional impl blocks: one `for &T` and one
//! // `for Box<T>` where `T: Foo`.
//! #[auto_impl(&, Box)]
//! trait Foo {
//!     fn foo(&self);
//! }
//!
//! impl Foo for i32 {
//!     fn foo(&self) {}
//! }
//!
//! fn requires_foo(_: impl Foo) {}
//!
//!
//! requires_foo(0i32);  // works: through the impl we defined above
//! requires_foo(&0i32); // works: through the generated impl
//! requires_foo(Box::new(0i32)); // works: through the generated impl
//! ```
//!
//!
//! # Basic syntax and supported types
//!
//! You can annotate your trait with the `#[auto_impl(...)]` attribute. That
//! attribute can only be used on traits and not on structs, enums or anything
//! else.
//!
//! In the attribute, you have to specify all so called *proxy types* (the
//! types you want to generate impls for) as a comma separated list. Each proxy
//! type has a short abbreviation that you have to list there.
//!
//! Currently the following proxy types are supported:
//!
//! | Abbreviation | Example generated impl |
//! | ------------ | ---------------------- |
//! | `&`          | `impl<T: Trait> Trait for &T` |
//! | `&mut`       | `impl<T: Trait> Trait for &mut T` |
//! | `Box`        | `impl<T: Trait> Trait for Box<T>` |
//! | `Rc`         | `impl<T: Trait> Trait for Rc<T>` |
//! | `Arc`        | `impl<T: Trait> Trait for Arc<T>` |
//! | `Fn`         | `impl<T: Fn()> Trait for T` |
//! | `FnMut`      | `impl<T: FnMut()> Trait for T` |
//! | `FnOnce`     | `impl<T: FnOnce()> Trait for T` |
//!
//!
//! # More examples
//!
//! More examples can be found in [the examples folder][examples]. In
//! particular, the `greet_closure` example shows how to use the `Fn*` proxy
//! types.
//!
//! [examples]: https://github.com/auto-impl-rs/auto_impl/tree/master/examples
//!
//! The following example shows that a trait can contain associated consts,
//! associated types and complex methods (with generics, bounds, ...).
//!
//! ```
//! use auto_impl::auto_impl;
//! use std::{fmt, rc::Rc};
//!
//!
//! #[auto_impl(&, &mut, Box, Rc)]
//! trait Animal {
//!     const NUMBER_OF_LEGS: u8;
//!
//!     type Name: fmt::Display;
//!     fn name(&self) -> Self::Name;
//!
//!     fn select_favorite<'a, I>(&self, toys: I) -> &'a str
//!     where
//!         I: Iterator<Item = &'a str>;
//! }
//!
//! struct Dog(String);
//!
//! impl Animal for Dog {
//!     const NUMBER_OF_LEGS: u8 = 4;
//!
//!     type Name = String;
//!     fn name(&self) -> Self::Name {
//!         self.0.clone()
//!     }
//!
//!     fn select_favorite<'a, I>(&self, mut toys: I) -> &'a str
//!     where
//!         I: Iterator<Item = &'a str>
//!     {
//!         toys.next().unwrap()
//!     }
//! }
//!
//! fn require_animal(_: impl Animal) {}
//!
//! // All these calls work, as the `#[auto_impl]` attribute generated four
//! // impls for all those proxy types
//! require_animal(Dog("Doggo".into()));
//! require_animal(&Dog("Doggo".into()));
//! require_animal(&mut Dog("Doggo".into()));
//! require_animal(Box::new(Dog("Doggo".into())));
//! require_animal(Rc::new(Dog("Doggo".into())));
//! ```
//!
//!
//! # Restriction of references and smart pointers
//!
//! Not every trait can be implemented for every proxy type. As an easy
//! example, consider this trait:
//!
//! ```
//! trait Bar {
//!     fn bar(&mut self);
//! }
//! ```
//!
//! If we try to implement it for immutable references via `#[auto_impl(&)]`
//! the following impl would be generated:
//!
//! ```ignore
//! impl<T: Bar> Bar for &T {
//!     fn bar(&mut self) {
//!         T::bar(*self)  // fails to compile
//!     }
//! }
//! ```
//!
//! As you can easily see, this won't work because we can't call `bar` through
//! an immutable reference. There are similar restrictions for many other
//! smart pointers and references.
//!
//! In the following table you can see which methods can be implemented for
//! which proxy type. If a trait contains at least one method that cannot be
//! implemented for a proxy type, you cannot implement the trait for that proxy
//! type.
//!
//! | Trait contains method with... | `&` | `&mut` | `Box` | `Rc` | `Arc` |
//! | ----------------------------- | --- | ------ | ----- | ---- | ----- |
//! | `&self` receiver              | ✔   | ✔      | ✔     | ✔    | ✔     |
//! | `&mut self` receiver          | ✗   | ✔      | ✔     | ✗    | ✗     |
//! | `self` receiver               | ✗   | ✗      | ✔     | ✗    | ✗     |
//! | no `self` receiver            | ✔   | ✔      | ✔     | ✔    | ✔     |
//!
//! References and smart pointers have **no restriction in regard to associated
//! types and associated consts**! Meaning: traits with associated types/consts
//! can always be implemented for references and smart pointers as long as the
//! methods of that trait can be implemented.
//!
//!
//! # Restriction of closure types (`Fn*` traits)
//!
//! The `Fn*` proxy types have a lot more restrictions than references and
//! smart pointer:
//! - the trait must not define any associated types or consts
//! - the trait must define **exactly one** method
//!     - the method must have a `self` receiver
//!     - the method must not return anything borrowed from `self`
//!     - the method must not have generic type or const parameters
//!
//! Additionally, some `Fn*` traits cannot be implemented for all `self`
//! receiver types:
//!
//! | `self` Receiver | `Fn` | `FnMut` | `FnOnce` |
//! | --------------- | ---- | ------- | -------- |
//! | `&self`         | ✔    | ✗       | ✗        |
//! | `&mut self`     | ✔    | ✔       | ✗        |
//! | `self`          | ✔    | ✔       | ✔        |
//!
//! Lastly, the impls generated for the `Fn*` proxy types contain `for T`. This
//! is the most general blanket impl. So just be aware of the problems with
//! coherence and orphan rules that can emerge due to this impl.
//!
//!
//! # The `keep_default_for` attribute for methods
//!
//! By default, the impls generated by `auto_impl` will overwrite all methods
//! of the trait, even those with default implementation. Sometimes, you want
//! to not overwrite default methods and instead use the default
//! implementation. You can do that by adding the
//! `#[auto_impl(keep_default_for(...))]` attribute to a default method. In the
//! parenthesis you need to list all proxy types for which the default method
//! should be kept.
//!
//! From [the `keep_default_for` example](
//! https://github.com/auto-impl-rs/auto_impl/blob/master/examples/keep_default_for.rs):
//!
//! ```
//! # use auto_impl::auto_impl;
//! #[auto_impl(&, Box)]
//! trait Foo {
//!     fn required(&self) -> String;
//!
//!     // The generated impl for `&T` will not override this method.
//!     #[auto_impl(keep_default_for(&))]
//!     fn provided(&self) {
//!         println!("Hello {}", self.required());
//!     }
//! }
//! ```

extern crate proc_macro;
#[macro_use]
extern crate quote;

use proc_macro::TokenStream;

mod analyze;
mod attr;
mod gen;
mod proxy;

/// See crate documentation for more information.
#[proc_macro_attribute]
pub fn auto_impl(args: TokenStream, input: TokenStream) -> TokenStream {
    match auto_impl2(args, input.into()) {
        Ok(tokens) => tokens.into(),
        Err(e) => e.into_compile_error().into(),
    }
}

fn auto_impl2(
    args: TokenStream,
    input: proc_macro2::TokenStream,
) -> syn::Result<proc_macro2::TokenStream> {
    // Try to parse the token stream from the attribute to get a list of proxy
    // types.
    let proxy_types = proxy::parse_types(args);

    let mut trait_def = syn::parse2::<syn::ItemTrait>(input)?;

    let generated = gen::gen_impls(&proxy_types, &trait_def)?;

    // Before returning the trait definition, we have to remove all
    // `#[auto_impl(...)]` attributes on all methods.
    attr::remove_our_attrs(&mut trait_def)?;

    Ok(quote!(#trait_def #generated))
}
