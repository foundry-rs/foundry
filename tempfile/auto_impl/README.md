# `auto_impl` [![CI](https://github.com/auto-impl-rs/auto_impl/actions/workflows/ci.yml/badge.svg)](https://github.com/auto-impl-rs/auto_impl/actions/workflows/ci.yml) [![Crates.io](https://img.shields.io/crates/v/auto_impl.svg)](https://crates.io/crates/auto_impl) [![docs](https://docs.rs/auto_impl/badge.svg)](https://docs.rs/auto_impl)

A proc-macro attribute for automatically implementing a trait for references,
some common smart pointers and closures.

# Usage

This library requires Rust 1.56.0 or newer. This library doesn't leave any public API in your code.

Add `auto_impl` to your `Cargo.toml` and just use it in your crate:

```rust
// In Rust 2015 you still need `extern crate auto_impl;` at your crate root
use auto_impl::auto_impl;
```

Add an `auto_impl` attribute to traits you want to automatically implement for wrapper types. Here is a small example:

```rust
// This will generate two additional impl blocks: one for `&T` and one
// for `Box<T>` where `T: Foo`.
#[auto_impl(&, Box)]
trait Foo {
    fn foo(&self);
}

impl Foo for i32 {
    fn foo(&self) {}
}

fn requires_foo(_: impl Foo) {}


requires_foo(0i32);  // works: through the impl we defined above
requires_foo(&0i32); // works: through the generated impl
requires_foo(Box::new(0i32)); // works: through the generated impl
```

For more explanations, please see [**the documentation**](https://docs.rs/auto_impl) and for more examples, see 
[the examples folder](https://github.com/auto-impl-rs/auto_impl/tree/master/examples).

# Alternatives

This library implements a fraction of a very broad and complex usecase. It's mostly useful for applications that 
define traits for components, and want to be able to abstract over the storage for those traits. If it doesn't offer 
some functionality you need, check out the [`impl-tools`](https://github.com/kas-gui/impl-tools/) project.

---

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.
