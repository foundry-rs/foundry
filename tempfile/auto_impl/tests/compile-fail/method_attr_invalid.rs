use auto_impl::auto_impl;


#[auto_impl(&)]
trait Foo {
    #[auto_impl(ferris_for_life)]
    fn a(&self);
}


fn main() {}
