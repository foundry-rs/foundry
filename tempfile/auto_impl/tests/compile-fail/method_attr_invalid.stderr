error: expected arguments for 'ferris_for_life' in parenthesis `()`, found nothing
 --> $DIR/method_attr_invalid.rs:6:17
  |
6 |     #[auto_impl(ferris_for_life)]
  |                 ^^^^^^^^^^^^^^^
